### DISCUSSION TITLE
How to introduce Modelica.Constants in classes

### DISCUSSION THREAD

**Discussion by tobolar** [modelica/ModelicaStandardLibrary#3941] [Q&A] [answered]
There are several possibilities how to introduce a constant from __Modelica.Constants__ in a class. E.g. for `pi`, I have found following in MSL: 
- Definition of constant: `constant Real pi = Modelica.Constants.pi;` (or also `constant SI.Angle pi = Modelica.Constants.pi;`)
- Qualified import: `Import Modelica.Constants.pi;`
- Directly in equation, e.g. `phi = 2 * Modelica.Constants.pi * w;`

Is there a preferrable way having advantages over others?

**Comment by HansOlsson** [comment]
I would suggest avoiding the first one (new constant). I would prefer `Modelica.Constants.pi` (or `.Modelica.Constants.pi`); unless it is used a lot of times and the import would make it easier to reader. The problem with import is that you always have to check that it is imported in the current class.

**Reply by tobolar** [reply]
> The problem with import is that you always have to check that it is imported in the current class.

This is probably less problematic for common constants, compared to other classes. I mean when there is used `pi` somewhere in an equation, one can expect it could be an imported variable.

**Comment by thorade** [comment, marked as answer]
One more variant: 
`final constant pi = Modelica.Constants.pi` because otherwise the constant could be modified with a modifier from an extending class.
